ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/state/cache/PendingBlockCache.java

```diff
@@ -34,25 +37,39 @@ public PendingBlockCache(final long cacheSizePerPeer) {
    *
    * @return the previous value associated with the specified key, or {@code null} if there was no
    *     mapping for the hash
-   * @throws IndexOutOfBoundsException if the limit of the number of blocks has been reached for
-   *     this nodeId
    */
   @Override
   public ImmutablePendingBlock putIfAbsent(
-      final Hash hash, final ImmutablePendingBlock pendingBlock) throws IndexOutOfBoundsException {
-    if (getPeerWeight(pendingBlock.nodeId()) >= cacheSizePerPeer) {
-      throw new IndexOutOfBoundsException();
+      final Hash hash, final ImmutablePendingBlock pendingBlock) {
+    final ImmutablePendingBlock foundBlock = super.putIfAbsent(hash, pendingBlock);
+    if (foundBlock == null) {
+      removeLowestPriorityBlockWhenCacheFull(pendingBlock.nodeId());
     }
-    return super.putIfAbsent(hash, pendingBlock);
+    return foundBlock;
   }
 
   /**
-   * Returns the number of pending blocks from a node that are stored in the cache
+   * Removes the lowest priority block if a peer has reached the cache limit it is allowed to use
+   * The highest priority blocks are those that are lowest in block height and then higher priority
+   * if they were sent more recently.
    *
-   * @param nodeId the peer ID
-   * @return the number of elements in the cache coming from this node
+   * @param nodeId id of the peer
    */
-  private long getPeerWeight(final Bytes nodeId) {
-    return values().stream().filter(value -> value.nodeId() == nodeId).count();
+  private void removeLowestPriorityBlockWhenCacheFull(final Bytes nodeId) {
+    final List<ImmutablePendingBlock> blockByNodeId =
+        values().stream().filter(value -> value.nodeId() == nodeId).collect(Collectors.toList());
+    if (blockByNodeId.size() > cacheSizePerPeer) {
+      blockByNodeId.stream()
+          .min(getComparatorByBlockNumber().reversed().thenComparing(getComparatorByTimeStamp()))
```

Comment:
(optional) Regardless of the outcome of the other thread, if you want you can also represent this with stream operations:
```
    values().stream()
        .filter(value -> value.nodeId() == nodeId)
        .sorted(comparatorThatHasTheHighestPriorityFirst)
        .skip(cacheSizePerPeer)
        .forEach(value -> remove(value.block().getHash()));
```